experimental/StandardFiniteFields/src/StandardFiniteFields.jl

```diff
@@ -0,0 +1,509 @@
+  ext = get_standard_extensions!(F)
+  get!(f, ext, k)
+end
+
+
+function standard_affine_shift_data(q::IntegerUnion)
+  m = div(4 * q, 5)
+  while gcd(m, q) != 1
+    m -= 1
+  end
+  a = div(2 * q, 3)
+  return m, a
+end
+
+function standard_affine_shift(q::IntegerUnion, i::IntegerUnion)
+  m, a = standard_affine_shift_data(q)
+  return mod((m * i + a), q)
+end
+
+# Given a standard finite field F and Steinitz number n,
+# give the corresponding field element.
+function element_from_steinitz_number(F::PrimeField, n::IntegerUnion)
+  @req 0 <= n <= order(F) "We need to have 0 <= n <= q"
+  return F(n)
+end
+function element_from_steinitz_number(F::FinField, n::IntegerUnion)
+  @req is_standard_finite_field(F) "First input must be a standard finite field"
+  p = characteristic(F)
+  @req 0 <= n <= order(F) "We need to have 0 <= n <= q"
+  n == 0 && return zero(F)
+
+  # this forms a linear combo of F.towervasis rows using vectorrep as coefficients,
+  # and then convert this vector to an element of F.
+  vectorrep = digits(n, base = Int(p))
+  return F(vectorrep * @view tower_basis(F)[1:length(vectorrep), :])
+end
+
+# Returns an element a in F that is NOT an rth root of unity
+# we REQUIRE that F is a standard finite field
+function non_rth_root(F::FinField, r::IntegerUnion)
+  @assert is_standard_finite_field(F) || is_standard_prime_field(F)
+  q = order(F)
+  if mod(q - 1, r) == 0
+    i = 0
+    a = zero(F)
+    k = div(q - 1, r)
+    while iszero(a) || isone(a^k)
+      i += 1
+      a = element_from_steinitz_number(F, standard_affine_shift(q, i))
+    end
+    return a
+  else
+    return nothing
+  end
+end
+
+function standard_irreducible_coefficient_list(
+  F::FinField,
+  r::IntegerUnion,
+  a::FinFieldElem,
+)
+  q = order(F)
+  l = zeros(F, Int(r) + 1)
+  l[Int(r)+1] = one(F)
+  l[1] = a
+  l[2] = one(F)
+  # inc is the expected number of nonzero coefficients
+  inc = 1
+  let t = q
+    while t < 2 * r
+      t *= q
+      inc += 1
+    end
+  end
+  # allowing non-zero coeffs up to position d
+  # after every r attempts allow inc more non-zero coeffs
+  d = 0
+  count = 0
+  qq = 0
+  while !is_irreducible(polynomial(F, l))
+    if mod(count, r) == 0 && d < r - 1
+      d += inc
+      if d >= r
+        d = r - 1
+      end
+      qq = q^(d - 1)
+    end
+    # q can be very very large so Int is not big enough...
+    st = digits(standard_affine_shift(qq, count), base = BigInt(q), pad = d - 1)
+    # TODO: we can remove this while loop when fix for padding digits live
+    while length(st) < d - 1
+      push!(st, 0)
+    end
+    for k = 2:d
+      l[k] = element_from_steinitz_number(F, st[k-1])
+    end
+    count += 1
+  end
+  return l
+end
+
+# returns the Steinitz number corresponding to the polynomial g(X),
+# where f = X^r + g(X) is the standard irreducible polynomial over FF(p, r^(k-1))
+function steinitz_number_for_prime_degree(
+  p::IntegerUnion,
+  r::IntegerUnion,
+  k::IntegerUnion,
+)::ZZRingElem
```

Comment:
Return type annotation should, in general, not be used. See https://docs.julialang.org/en/v1/manual/functions/#Return-type (last paragraph)

Reply:
Hm, I've tried to avoid it in general. But in this case, I have several possible return cases, and I want to make sure that no matter what the result is converted and returned as a `ZZRingElem` and not an `Integer`.... I actually think I can get around this by making sure that `steinitz_number` will always return a `ZZRingElem`.

Reply:
Great! Explicit conversion in each of the return cases is the way to go. You can check if you got all cases with `@code_warntype`